src/transformers/models/falcon/modeling_falcon.py

```diff
@@ -209,6 +209,162 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
+        return max(low, 0), min(high, dim - 1)
+
+    def linear_ramp_mask(self, min, max, dim):
+        if min == max:
+            max += 0.001  # Prevent singularity
+
+        linear_func = (torch.arange(dim, dtype=torch.float32) - min) / (max - min)
+        ramp_func = torch.clamp(linear_func, 0, 1)
+        return ramp_func
+
+    def get_mscale(self, scaling_factor=1):
+        if scaling_factor <= 1:
+            return 1.0
+        return 0.1 * math.log(scaling_factor) + 1.0
+
+    def forward(self, x, seq_len=None):
+        # Difference to the original RoPE: applies a scaling factor computed with
+        # the YaRN method (NTK-by-Parts + Attn Scaling)
+        # x: [batch_size, seq_len, head_dim]
+        if seq_len > self.max_seq_len_cached:
+            self.max_seq_len_cached = seq_len
+            emb = self.get_pos_embeddings(x.device)
+
+            self._cos_cached = (emb.cos() * self.mscale)[:, :].to(x.dtype)
+            self._sin_cached = (emb.sin() * self.mscale)[:, :].to(x.dtype)
+        return (
+            self._cos_cached[:seq_len, ...].to(dtype=x.dtype),
+            self._sin_cached[:seq_len, ...].to(dtype=x.dtype),
+        )
+
+    def yarn(self, device):
+        pos_freqs = self.base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim)
+        inv_freq_extrapolation = 1.0 / pos_freqs
+        inv_freq_interpolation = 1.0 / (self.scaling_factor * pos_freqs)
+
+        low, high = self.find_correction_range(
+            self.beta_fast, self.beta_slow, self.dim, self.base, self.original_max_position_embeddings
+        )
+        # Get n-dimensional rotational scaling corrected for extrapolation
+        inv_freq_mask = (
+            1 - self.linear_ramp_mask(low, high, self.dim // 2).float().to(device)
+        ) * self.extrapolation_factor
+        inv_freq = inv_freq_interpolation * (1 - inv_freq_mask) + inv_freq_extrapolation * inv_freq_mask
+
+        self.register_buffer("inv_freq", inv_freq)
+        # Get n-dimensional magnitude scaling corrected for interpolation
+        self.mscale = float(self.get_mscale(self.scaling_factor) * self.attention_factor)
+
+
+class FalconDynamicYaRNScalingRotaryEmbedding(FalconYaRNScalingRotaryEmbedding):
+    def __init__(
+        self,
+        dim,
+        max_position_embeddings=2048,
+        base=10000,
+        scaling_factor=1,
+        original_max_position_embeddings=2048,
+        extrapolation_factor=1,
+        attention_factor=1,
+        beta_fast=32,
+        beta_slow=1,
+        finetuned=False,
+        device=None,
+    ):
+        super().__init__(
+            dim,
+            max_position_embeddings,
+            base,
+            scaling_factor,
+            original_max_position_embeddings,
+            extrapolation_factor,
+            attention_factor,
+            beta_fast,
+            beta_slow,
+            device,
+        )
+
+        if finetuned:
```

Comment:
```suggestion
        if self.max_position_embeddings != self.original_max_position_embeddings:
```
This should be true for fine-tuned models, saving us a flag :)